### PR TITLE
bugfix: rmhc masking

### PR DESCRIPTION
**Pull request by koritsky**
https://github.com/yaak-ai/rmind/pull/160 introduced a bug when masking wasn't applied to `episode.input_embeddings` in `RandomMaskedHindsightControl`. In a nutshell, `TensorDict` handles indexes and slices differently in `.apply_()`. Hence it was replaced with `.masked_apply_()`

Interestingly enough, such changes are not necessary at all with current setup (so simple `= 1.0` is sufficient). L2 normalization in https://github.com/yaak-ai/rmind/pull/156 converts `float16` to `float32` during episode building removing the necessity to consider `dtypes` of `episode.input_embeddings`. However, this is pure luck and proper tensor dtypes handing from this PR should be implemented.

https://wandb.ai/yaak/cargpt/runs/p3szi3o0/overview

[sc-34148]

### CODE REVIEW

File: src/rmind/components/objectives/random_masked_hindsight_control.py [hunk 1]
```diff
@@ -46,19 +46,34 @@ def __init__(
...
         masked_action_timestep_idx = np.random.choice(t, 2, replace=False).tolist()  # noqa: NPY002
         masked_observation_timestep_idx = np.random.choice(t, 1, replace=False).tolist()  # noqa: NPY002
 
         episode = episode.clone(recurse=True)
         episode.input_embeddings.select(
             *episode.timestep.keys_by_type[TokenType.ACTION]
-        )[:, masked_action_timestep_idx].apply_(lambda x: x.fill_(-1.0))
+        ).masked_fill_(
+            torch.zeros((b, t), dtype=torch.bool, device=device).index_fill_(
+                1,
```
Comment (sandhawalia) [inline src/rmind/components/objectives/random_masked_hindsight_control.py:60]: for clarity of understanding 

```
# build mask once
mask = torch.zeros((b, t), dtype=torch.bool, device=device)
mask[:, masked_action_timestep_idx] = True

episode.input_embeddings.select(
            *episode.timestep.keys_by_type[TokenType.OBSERVATION]
        ).masked_fill_(mask, -1.0)

episode.input_embeddings.select(
            *episode.timestep.keys_by_type[TokenType.ACTION]
        ).masked_fill_(mask, -1.0)

```

File: src/rmind/components/objectives/random_masked_hindsight_control.py [hunk 2]
```diff
@@ -110,11 +126,25 @@ def predict(
...
+                    True,  # noqa: FBT003
+                ),
+                -1.0,
+            )
 
             episode.input_embeddings.select(
                 *episode.timestep.keys_by_type[TokenType.OBSERVATION]
-            )[:, masked_observation_timestep_idx] = -1.0
+            ).masked_fill_(
+                torch.zeros((b, t), dtype=torch.bool, device=device).index_fill_(
```
Comment (sandhawalia) [inline src/rmind/components/objectives/random_masked_hindsight_control.py:141]: likewise as above

**Review by sandhawalia — APPROVED** [review]
Small change needed for clarity. separating mask and assignment

File: src/rmind/components/objectives/random_masked_hindsight_control.py (hunk 1 above)
Comment (koritsky) [reply src/rmind/components/objectives/random_masked_hindsight_control.py:60]: can't do: indexes are different for action and observation
